extensions/src/Smithy.Identity.Abstractions/Auth/IAuthSchemeResolver.cs

```diff
@@ -0,0 +1,31 @@
+﻿using System.Collections.Generic;
+
+namespace Smithy.Identity.Abstractions
+{
+    /// <summary>
+    /// <para>
+    /// An authentication scheme resolver is an entity within the SDK that determines which authentication scheme to use 
+    /// for a service.
+    /// </para>
+    /// 
+    /// <para>
+    /// It accepts data from the request (e.g. request members) and client (e.g. AWS region) and returns a list of
+    /// authentication scheme options that the SDK should use when authenticating that request.
+    /// </para>
+    /// </summary>
+    public interface IAuthSchemeResolver
```

Comment:
Do we need to have this be a generic interface where the generic is the type of `IAuthSchemeParameters`? and then use the `T` in the `ResolveAuthScheme`? Otherwise I'm not sure how in the auth resolver which will need the service specific auth parameters without doing an awkward casting.

```
public interface IAuthSchemeResolver<T> : where T : IAuthSchemeParameters
{
    List<IAuthSchemeOption> ResolveAuthScheme(T authParameters);
}
```

Reply:
Refactored this interface in https://github.com/aws/aws-sdk-net/pull/3382/commits/e6ed5012726d5b8cbd6c7d1c0c8838c3bfc55302

Reply:
That was a good suggestion, the service resolver will be easier to implement (no need to cast anything):
```csharp
public class AutoScalingAuthSchemeParameters : IAuthSchemeParameters
{
    public string Operation { get; set; }
    public string Region { get; set; }
}

public class AutoScalingAuthSchemeResolver : IAuthSchemeResolver<AutoScalingAuthSchemeParameters>
{
    public List<IAuthSchemeOption> ResolveAuthScheme(AutoScalingAuthSchemeParameters authParameters)
    {
        // Can read region here.
        Console.WriteLine(authParameters.Region);
    }
}
```